Add LogicalPlan::SubqueryAlias

andygrove opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #https://github.com/apache/arrow-datafusion/issues/2164.

 # Rationale for this change
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

I would like the logical plan to better represent aliased relations to help with translation to a different execution engine. Currently, original table names are discarded during SQL planning.

# What changes are included in this PR?
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

Introduces new `SubqueryAlias` node in `LogicalPlan` and updates the SQL planner to wrap this around `TableScan` rather than just use the alias name in the `TableScan`.

# Are there any user-facing changes?
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

The query plan will look different now.

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

alamb commented on datafusion/core/src/logical_plan/builder.rs:
```diff
@@ -518,6 +518,18 @@ impl LogicalPlanBuilder {
     }
 
+    /// Apply an alias
+    pub fn alias(&self, alias: &str) -> Result<Self> {
```
Ah, and I see you don't want to do that :) in the description

I wonder if you could use the `DFSchema::relation` name for this purpose?

andygrove replied:
I think I need something to very specifically say that a table is being used as an alias. An easier change might be just to add an additional field to the `TableScan` to record the original table name. I think I will put up a separate PR for that approach.

So to explain why I need this. I want to use DataFusion as a SQL parser and planner but I want to execute the query in a different engine. I can provide a `TableProvider` so that DataFuision can get the schema for table `person` and I get a logical plan. When I go to translate that plan to a physical plan, it refers to a table called `peeps` (the alias) and I have no way to know the actual table name.

andygrove replied:
This is the Spark plan for this use case:

```
'Project [*]
+- 'SubqueryAlias peeps
   +- 'UnresolvedRelation [person], [], false
```

Dandandan replied:
I think I prefer the approach taken in this PR, as it could handle subqueries rather than just table scans (as suggested by the name).

I think this is a cleaner, more generic approach.

mingmwang commented:
Looks like in DataFusion, the logical Project node already support Alias. It should be similar to SparkSQL's SubqueryAlias.
And we can also use it to represent a NamedStruct

````

pub struct Projection {
    /// The list of expressions
    pub expr: Vec<Expr>,
    /// The incoming logical plan
    pub input: Arc<LogicalPlan>,
    /// The schema description of the output
    pub schema: DFSchemaRef,
    /// Projection output relation alias
    pub alias: Option<String>,
}


````

andygrove commented:
@alamb @Dandandan @mingmwang I have fixed the test failures and renamed `AliasedRelation` to `SubqueryAlias`. This is only implemented for `TableScan` so far but if this approach is acceptable, I will file follow-on issues to use the same approach for projection and union, which both currently have their own `alias` fields.

Dandandan commented on datafusion/core/src/sql/planner.rs:
```diff
@@ -644,16 +644,15 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
-                            // take alias into account to support `JOIN table1 as table2`
-                            alias
-                                .as_ref()
-                                .map(|a| a.name.value.as_str())
```
👍

alamb commented on datafusion/core/src/optimizer/projection_push_down.rs:
```diff
@@ -432,6 +432,34 @@ fn optimize_plan(
                 alias: alias.clone(),
             }))
         }
+        LogicalPlan::SubqueryAlias(SubqueryAlias { input, alias, .. }) => {
```
is it worth a test for projection pushdown specifically?

andygrove replied:
Good point. Test added.

alamb commented on datafusion/core/src/sql/planner.rs:
```diff
@@ -3458,7 +3458,8 @@ mod tests {
         \n  Inner Join: Using #person.id = #person2.id\
         \n    TableScan: person projection=None\
-        \n    TableScan: person2 projection=None";
+        \n    SubqueryAlias: person2\
```
the appearance of `SubqueryAlias` in this query is  somewhat strange to me as there is no subquery in this SQL (maybe a better name is `RelationAlias` or something)?

andygrove replied:
I originally named this `AliasedRelation` and then changed it to `SubqueryAlias` because that is what Spark uses. I don't have a strong preference either way.

alamb replied:
me neither. Whatever you prefer is fine with me

alamb reviewed (approved):
Seems like an improvement to me -- I think the projection pushdown code might benefit from an additional test, but I also think the existing coverage is likely sufficient.

alamb commented:
Thanks @andygrove
